File: src/commands/add-note.ts
import fs from 'node:fs/promises';
import { parseArgs } from 'node:util';

import { Ticket } from '../lib/ticket.ts';
import { isoDate } from '../lib/utils.ts';
import { exitWithError, readStdin } from './utils.ts';

export async function cmdAddNote(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 1) {
    exitWithError('Usage: tk add-note <id> [note text]');
  }

  const ticket = await Ticket.find(positionals[0]);
  let note = positionals.slice(1).join(' ') || '';

  if (!process.stdin.isTTY) {
    note = await readStdin();
  }

  let content = await fs.readFile(ticket.filePath, 'utf-8');

  if (!content.includes('## Notes')) {
    content += '\n## Notes';
  }

  const timestamp = isoDate();
  content += `\n\n**${timestamp}**\n\n${note}`;

  await fs.writeFile(ticket.filePath, content, 'utf-8');
  console.log(`Note added to ${ticket.id}`);
}

File: src/commands/add_note.rs
use serde_json::json;

use super::CommandOutput;
use crate::error::{JanusError, Result};
use crate::events::log_note_added;
use crate::ticket::Ticket;
use crate::utils::{is_stdin_tty, iso_date, read_stdin};

/// Add a timestamped note to a ticket
pub async fn cmd_add_note(id: &str, note_text: Option<&str>, output_json: bool) -> Result<()> {
    let ticket = Ticket::find(id).await?;

    // Get note text from argument or stdin
    let note = if let Some(text) = note_text {
        text.to_string()
    } else if !is_stdin_tty() {
        read_stdin()?
    } else {
        String::new()
    };

    // Validate that note is not empty or only whitespace
    if note.trim().is_empty() {
        return Err(JanusError::EmptyNote);
    }

    let timestamp = iso_date();

    let content = ticket.read_content()?;
    let mut new_content = content;
    if !new_content.contains("## Notes") {
        new_content.push_str("\n## Notes");
    }
    new_content.push_str(&format!("\n\n**{}**\n\n{}", timestamp, note));
    ticket.write(&new_content)?;

    log_note_added(&ticket.id, &note);

    CommandOutput::new(json!({
        "id": ticket.id,
        "action": "note_added",
        "timestamp": timestamp,
        "note": note,
    }))
    .with_text(format!("Note added to {}", ticket.id))
    .print(output_json)
}
